pecos/core/utils/mmap_util.hpp

```diff
@@ -0,0 +1,509 @@
+         * If open for read, should NOT close until the data of the file is not visited anymore.
+         */
+        void close() {
+            if (_mode == _Mode::UNINIT) {
+                return;
+            }
+            else if (_mode == _Mode::READONLY) {
+                delete _mmap_r;
+                _mmap_r = nullptr;
+            }
+            else if (_mode == _Mode::WRITEONLY) {
+                delete _mmap_w;
+                _mmap_w = nullptr;
+            }
+            _mode = _Mode::UNINIT;
+        }
+
+        /* Expose get/put functions */
+        template<class T>
+        T fget_one() {
+            _check_get();
+            return _mmap_r->fget_one<T>();
+        }
+
+        template<class T>
+        T* fget_multiple(const uint64_t n_elements) {
+            _check_get();
+            return _mmap_r->fget_multiple<T>(n_elements);
+        }
+
+        template<class T>
+        void fput_one(const T& src) {
+            _check_put();
+            _mmap_w->fput_one<T>(src);
+        }
+
+        template<class T>
+        void fput_multiple(const T * src, const uint64_t n_elements) {
+            _check_put();
+            _mmap_w->fput_multiple<T>(src, n_elements);
+        }
+
+    private:
+        enum class _Mode {
+            UNINIT = 0,             // Uninitialized
+            READONLY = 1,           // "r"
+            WRITEONLY = 2,          // "w"
+        };
+
+        _details::_MmapFileSave * _mmap_w = nullptr; // Used for write mode
+        _details::_MmapFileLoad * _mmap_r = nullptr; // Used for read mode
+        _Mode _mode = _Mode::UNINIT; // Indicator for current opened mode
+
+        /* Check if opened as read mode for calling get functions */
+        void _check_get() {
+            if (!_mmap_r) {
+                throw std::runtime_error("Not opened for read mode, cannot call get.");
+            }
+        }
+
+        /* Check if opened as write mode for calling put functions */
+        void _check_put() {
+            if (!_mmap_w) {
+                throw std::runtime_error("Not opened for write mode, cannot call put.");
+            }
+        }
+}; // end class MmapFile
+
+
+/* Extended Vector class that can be used as a std::vector or as a memory-view from part of mmap file
+ * For std::vector case, it own the memory for data storage.
+ * For mmap view case, it does not own any memory, but serve as a view for a piece of memory owned by MmapFile.
+ * By default, it is initialized as empty std::vector that can be resized or loaded as mmap view.
+ * Once loaded as mmap view, it cannot go back to std::vector case unless clear() is called.
+ */
+template<class T, class TT = T, _details::if_simple_serializable<TT> = true>
+class MmapableVector {
+    public:
+        MmapableVector(uint64_t size=0) { resize(size); }
+
+        /* Functions to match std::vector interface */
+        uint64_t size() const { return _size; }
+
+        const T* data() const { return _data; }
```

Comment:
I think we might need both

const T* data() const { return _data; }
T* data() { return data; } 

?